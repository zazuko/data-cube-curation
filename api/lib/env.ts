import { warning } from './log'

const hasProxy = new Proxy(process.env, {
  get (env: Record<string, string>, prop: string) {
    return !!env[prop]
  },
})

const handler = {
  get (env: Record<string, string>, prop: string) {
    if (prop === 'has') {
      return hasProxy
    }

    const value = env[prop]

    if (!value && process.env.NODE_ENV === 'production') {
      warning(`Missing environment variable %s`, prop)
      return ''
    }

    return value
  },
}

export default new Proxy(process.env, handler) as Record<string, string> & {
  has: Record<string, boolean>;
}
