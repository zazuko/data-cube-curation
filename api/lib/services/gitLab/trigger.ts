import { URLSearchParams } from 'url'
import fetch from 'node-fetch'
import { Project } from '@zazuko/rdfine-data-cube'
import env from '../../env'

interface GitLabJobTrigger {
  s3Bucket?: string;
  graphUri?: string;
}

export function triggerPipeline (project: Pick<Project, 's3Bucket' | 'id' | 'graphUri'>, trigger: GitLabJobTrigger = {}) {
  const s3Bucket = trigger.s3Bucket || project.s3Bucket
  const graphUri = trigger.graphUri || project.graphUri

  if (!s3Bucket) {
    throw new Error('S3 bucket missing')
  }

  if (!graphUri) {
    throw new Error('Graph URI missing')
  }

  const form = new URLSearchParams({
    token: env.GITLAB_PIPELINE_TOKEN,
    ref: env.GITLAB_PIPELINE_BRANCH || 'master',
    'variables[PROJECT]': project.id.value,
    'variables[S3_ENDPOINT]': env.AWS_S3_ENDPOINT,
    'variables[S3_BUCKET]': s3Bucket,
    'variables[AWS_ACCESS_KEY_ID]': env.AWS_ACCESS_KEY_ID,
    'variables[AWS_SECRET_ACCESS_KEY]': env.AWS_SECRET_ACCESS_KEY,
    'variables[GRAPH_STORE_ENDPOINT]': env.GRAPH_STORE_ENDPOINT,
    'variables[GRAPH_STORE_USER]': env.GRAPH_STORE_USER,
    'variables[GRAPH_STORE_PASSWORD]': env.GRAPH_STORE_PASSWORD,
    'variables[GRAPH]': graphUri,
  })

  return fetch(env.GITLAB_PIPELINE, {
    method: 'POST',
    body: form,
  })
}
