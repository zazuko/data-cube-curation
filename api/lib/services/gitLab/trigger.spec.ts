import { namedNode } from 'rdf-data-model'
import { triggerPipeline } from './trigger'
import fetch from 'node-fetch'
import env from '../../env'

jest.mock('node-fetch')
jest.mock('../../env')

const envMock = env as any as jest.SpyInstance
const fetchMock = fetch as any as jest.SpyInstance

describe('gitlab trigger', () => {
  beforeAll(() => {
    env.GITLAB_PIPELINE = 'GITLAB_PIPELINE'
    env.GITLAB_PIPELINE_TOKEN = 'GITLAB_PIPELINE_TOKEN'
    env.GITLAB_PIPELINE_BRANCH = 'master'
    env.AWS_ACCESS_KEY_ID = 'AWS_ACCESS_KEY_ID'
    env.AWS_SECRET_ACCESS_KEY = 'AWS_SECRET_ACCESS_KEY'
    env.AWS_S3_ENDPOINT = 'AWS_S3_ENDPOINT'
  })

  beforeEach(() => {
    fetchMock.mockReset()
  })

  afterAll(() => {
    envMock.mockRestore()
  })

  it('throws if S3 bucket is not given', () => {
    // given
    const project = {
      id: namedNode('foo:bar'),
      s3Bucket: undefined,
    }

    // then
    expect(() => {
      // when
      triggerPipeline(project, {})
    }).toThrow()
  })

  it('calls gitlab trigger endpoint with project\'s S3 settings', async () => {
    // given
    const project = {
      id: namedNode('foo:bar'),
      s3Bucket: 'foo',
    }

    // when
    await triggerPipeline(project, {})

    // then
    expect(fetchMock.mock.calls[0]).toMatchSnapshot()
  })

  it('calls gitlab trigger endpoint with overridden S3 settings', async () => {
    // given
    const project = {
      id: namedNode('foo:bar'),
      s3Bucket: 'foo',
    }

    // when
    await triggerPipeline(project, {
      s3Bucket: 'changed bucket',
    })

    // then
    expect(fetchMock.mock.calls[0]).toMatchSnapshot()
  })
})
