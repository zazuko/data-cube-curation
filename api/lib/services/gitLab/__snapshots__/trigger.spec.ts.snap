// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gitlab trigger calls gitlab trigger endpoint with overridden S3 settings 1`] = `
Array [
  "GITLAB_PIPELINE",
  Object {
    "body": URLSearchParams {
      Symbol(query): Array [
        "token",
        "GITLAB_PIPELINE_TOKEN",
        "ref",
        "master",
        "variables[PROJECT]",
        "foo:bar",
        "variables[S3_ENDPOINT]",
        "AWS_S3_ENDPOINT",
        "variables[S3_BUCKET]",
        "changed bucket",
        "variables[AWS_ACCESS_KEY_ID]",
        "AWS_ACCESS_KEY_ID",
        "variables[AWS_SECRET_ACCESS_KEY]",
        "AWS_SECRET_ACCESS_KEY",
      ],
      Symbol(context): null,
    },
    "method": "POST",
  },
]
`;

exports[`gitlab trigger calls gitlab trigger endpoint with project's S3 settings 1`] = `
Array [
  "GITLAB_PIPELINE",
  Object {
    "body": URLSearchParams {
      Symbol(query): Array [
        "token",
        "GITLAB_PIPELINE_TOKEN",
        "ref",
        "master",
        "variables[PROJECT]",
        "foo:bar",
        "variables[S3_ENDPOINT]",
        "AWS_S3_ENDPOINT",
        "variables[S3_BUCKET]",
        "foo",
        "variables[AWS_ACCESS_KEY_ID]",
        "AWS_ACCESS_KEY_ID",
        "variables[AWS_SECRET_ACCESS_KEY]",
        "AWS_SECRET_ACCESS_KEY",
      ],
      Symbol(context): null,
    },
    "method": "POST",
  },
]
`;
